test/test_smalld.py

```diff
@@ -0,0 +1,158 @@
+from unittest.mock import Mock, patch
+
+import pytest
+from attrdict import AttrDict
+from smalld.smalld import GatewayClosedException, SmallD, recoverable_error_codes
+
+
+def prepare_gateway_mock(gateway_mock, smalld, side_effects=[()], auto_close=True):
+    """Configures __iter__ for gateway.
+
+    Each call to __iter__ is represented by an iterable in the side_effects list.
+    If the iterable yields an exception, the exception is raised and iteration halts.
+
+    By default this configuration:
+      - closes gracefully
+      - iterates once
+    """
+
+    side_effects = list(side_effects)
+    finished = False
+
+    def iter_gateway():
+        nonlocal finished
+        assert not finished, "infinite loop prevention"
+
+        if not side_effects:
+            finished = True
+            if auto_close:
+                smalld.close()
+            return
+
+        values = side_effects.pop(0)
+        for value in values:
+            if (
+                isinstance(value, BaseException)
+                or isinstance(value, type)
+                and issubclass(value, BaseException)
+            ):
+                raise value
+            yield AttrDict(value)
+
+    gateway_mock.__iter__.side_effect = iter_gateway
+
+
+@pytest.fixture(autouse=True)
+def gateway_mock():
+    with patch("smalld.smalld.Gateway") as gateway_cls:
+        yield gateway_cls.return_value
+
+
+@pytest.fixture(autouse=True)
+def client_mock():
+    with patch("smalld.smalld.HttpClient", autospec=True) as client_cls:
+        instance = client_cls.return_value
+        # for get gateway request
+        instance.get.return_value = AttrDict({"url": "url/to/gateway"})
+        yield instance
+
+
+@pytest.fixture(autouse=True)
+def sleep_mock():
+    with patch("time.sleep") as sleep_mock:
+        yield sleep_mock
+
+
+@pytest.fixture(autouse=True)
+def add_standard_listeners():
+    with patch("smalld.smalld.add_standard_listeners") as add_standard_listeners:
```

Comment:
If we didn't mock this, would it change the tests in any way?

Reply:
with the current tests, I don't think it would :thinking:; given that we always close smalld appropriately. letting side effects from listeners happen silently just seemed like a bad idea.

Reply:
In general, I think less mocks is good mocks.

Also, no reason why they have to let them happen silently. We could test them in the future if beneficial.